assets/js-text-with-image.js

```diff
@@ -3,79 +3,168 @@
-        // Replace the current palette with the one from the previous section
-        spacer.classList.replace(currentPalette, palette);
-      } else if (!currentPalette) {
+    let foundPrevPalette = null,
+        currentSpacerPalette = null;
+
+    const findPalette = (el) => {
+      for (let i = 0; i < config.class.palettes.length; i++) {
+        const palette = config.class.palettes[i];
+
+        if (hasPalette(el, palette)) return palette;
+      }
+      return null;
+    }
+
+    foundPrevPalette = findPalette(prevChild); // Find palettes in previous child and current spacer
+
+    if (!foundPrevPalette) return;
+
+    currentSpacerPalette = findPalette(spacer); // Find the current palette in the spacer
+
+    // Apply palette class efficiently using Utils with frame sequencing
+    const applyClasses = () => {
+      if (currentSpacerPalette && currentSpacerPalette !== foundPrevPalette) {
+        // Use frameSequence for better performance and transition handling
+        $.frameSequence(
+          // Read phase - no actual reads needed here, just prepare data
+          () => ({
+            remove: currentSpacerPalette,
+            add: foundPrevPalette
+          }),
+          // Write phase - perform DOM updates with proper sequencing
+          (data) => {
+            $.toggleClass(spacer, data.remove, false);
+            $.nextFrame(() => $.toggleClass(spacer, data.add, true));
+          }
+        )
+      } else if (!currentSpacerPalette) {
         // No palette on the spacer, just add the one from the previous section
-        spacer.classList.add(palette);
+        $.nextFrame(() => $.toggleClass(spacer, foundPrevPalette, true));
       }
-    })
+    }
+
+    $.batchDOM(applyClasses);
   }
 
   const setSpacerColors = () => {
-    if (!elements.prevSection) return;
+    const elements = getElements(),
+          prevSection = elements.prevSection;
 
-    const prevChildren = [...elements.prevSection.children];
-    if (!prevChildren.length) return;
+    if (!prevSection) return;
 
-    elements.children.forEach(child => {
-      if (!child.classList.contains(config.classes.spacer)) return;
+    const prevChildren = prevSection.children;
+    if (!prevChildren || prevChildren.length === 0) return;
 
-      // Apply palette from each element in previous section
-      prevChildren.forEach(prevChild => {
-        applyPaletteToSpacer(child, prevChild);
-      })
-    })
+    const children = elements.section.children,
+          spacers = [],
+          prevWithPalettes = [];
+
+    // Use a single function for the detection work
+    const findElements = () => {
+      for (let i = 0; i < children.length; i++) {
+        if (hasSpacerClass(children[i])) {
+          spacers.push(children[i]);
+        }
+      }
+
+      if (spacers.length === 0) return false;
+
+      // Find all previous children with palettes
+      for (let i = 0; i < prevChildren.length; i++) {
```

Comment:
can we use also `filter` here or `forEach`?. I don't like `for` in JS anymore xD